File: src/main/java/org/janelia/saalfeldlab/paintera/ui/source/state/StatePane.java
package org.janelia.saalfeldlab.paintera.ui.source.state;

import bdv.viewer.Source;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.beans.value.ObservableDoubleValue;
import javafx.scene.Node;
import javafx.scene.control.ButtonType;
import javafx.scene.control.CheckBox;
import javafx.scene.control.ContentDisplay;
import javafx.scene.control.Dialog;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.control.TitledPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import org.janelia.saalfeldlab.paintera.state.SourceInfo;
import org.janelia.saalfeldlab.paintera.state.SourceState;
import org.janelia.saalfeldlab.paintera.ui.BindUnbindAndNodeSupplier;
import org.janelia.saalfeldlab.paintera.ui.CloseButton;
import org.janelia.saalfeldlab.util.SciJavaUtils;
import org.scijava.InstantiableException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.invoke.MethodHandles;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;

public class StatePane implements BindUnbindAndNodeSupplier
{

	private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

	private static Map<Class<?>, SourceStateUIElementsFactory> UI_ELEMENTS_FACTORY_MAP = null;

	private static SourceStateUIElementsFactory getUiElementsFactory(Class<?> clazz)
	{
		if (UI_ELEMENTS_FACTORY_MAP == null)
		{
			final Map<Class<?>, SourceStateUIElementsFactory> tmp = new HashMap<>();
			try {
				SciJavaUtils.byTargetClassSortedByPriorities(SourceStateUIElementsFactory.class)
						.entrySet()
						.stream()
						.filter(Objects::nonNull)
						.filter(e -> e.getKey() != null && e.getValue() != null && !e.getValue().isEmpty())
						.forEach(e -> tmp.put(e.getKey(), e.getValue().get(0).getKey()));
			} catch (InstantiableException e) {
				throw new RuntimeException(e);
			}
			UI_ELEMENTS_FACTORY_MAP = new HashMap<>(tmp);
		}
		return UI_ELEMENTS_FACTORY_MAP.get(clazz);
	}

	private final SourceState<?, ?> state;

	private final SourceInfo sourceInfo;

	private final BindUnbindAndNodeSupplier supplier;

	private final TitledPane statePane;

	private final StringProperty name = new SimpleStringProperty();

	private final BooleanProperty isCurrentSource = new SimpleBooleanProperty();

	private final BooleanProperty isVisible = new SimpleBooleanProperty();

	public StatePane(
			final SourceState<?, ?> state,
			final SourceInfo sourceInfo,
			final Consumer<Source<?>> remove,
			final ObservableDoubleValue width)
	{
		super();
		this.state = state;
		this.sourceInfo = sourceInfo;
		this.supplier = Optional
				.ofNullable(getUiElementsFactory(state.getClass()))
				.orElseGet(SourceStateUIElementsDefaultFactory::new)
				.create(state);

		this.statePane = new TitledPane(null, this.supplier.get());
		this.statePane.minWidthProperty().bind(width);
		this.statePane.maxWidthProperty().bind(width);
		this.statePane.prefWidthProperty().bind(width);
		this.statePane.setExpanded(false);

		// create graphics for titled pane
		final Node closeButton = CloseButton.create(8);
		closeButton.setOnMousePressed(event -> remove.accept(state.getDataSource()));
		final Label sourceElementLabel = new Label(state.nameProperty().get(), closeButton);
		sourceElementLabel.textProperty().bind(this.name);
		sourceElementLabel.setOnMouseClicked(event -> {
			event.consume();
			if (event.getClickCount() != 2) { return; }
			final Dialog<Boolean> d = new Dialog<>();
			d.setTitle("Set source name");
			final TextField tf = new TextField(name.get());
			tf.setPromptText("source name");
			d.getDialogPane().getButtonTypes().addAll(ButtonType.OK, ButtonType.CANCEL);
			d.getDialogPane().lookupButton(ButtonType.OK).disableProperty().bind(tf.textProperty().isNull().or(tf
					.textProperty().length().isEqualTo(
					0)));
			d.setGraphic(tf);
			d.setResultConverter(ButtonType.OK::equals);
			final Optional<Boolean> result = d.showAndWait();
			if (result.isPresent() && result.get())
			{
				name.set(tf.getText());
			}
		});
		sourceElementLabel.setContentDisplay(ContentDisplay.RIGHT);
		sourceElementLabel.underlineProperty().bind(isCurrentSource);

		final HBox sourceElementButtons = getPaneGraphics(isVisible);
		sourceElementButtons.setMaxWidth(Double.MAX_VALUE);
		HBox.setHgrow(sourceElementButtons, Priority.ALWAYS);
		final HBox graphic = new HBox(sourceElementButtons, sourceElementLabel);
		graphic.setSpacing(20);
		//		graphic.prefWidthProperty().bind( this.width.multiply( 0.8 ) );
		this.statePane.setGraphic(graphic);
		//		addDragAndDropListener( statePane, this.info, contents.getChildren() );
	}

	@Override
	public Node get()
	{
		return this.statePane;
	}

	@Override
	public void bind()
	{
		this.name.bindBidirectional(state.nameProperty());
		this.isVisible.bindBidirectional(state.isVisibleProperty());
		this.isCurrentSource.bind(sourceInfo.isCurrentSource(state.getDataSource()));
		this.supplier.bind();
	}

	@Override
	public void unbind()
	{
		this.name.unbindBidirectional(state.nameProperty());
		this.isVisible.unbindBidirectional(state.isVisibleProperty());
		this.isCurrentSource.unbind();
		this.supplier.unbind();
	}

	private static HBox getPaneGraphics(final BooleanProperty isVisible)
	{
		final CheckBox cb = new CheckBox();
		cb.setMaxWidth(20);
		cb.selectedProperty().bindBidirectional(isVisible);
		cb.selectedProperty().set(isVisible.get());
		final HBox tp = new HBox(cb);
		return tp;
	}

}

File: src/main/kotlin/org/janelia/saalfeldlab/paintera/ui/source/state/StatePane.kt
package org.janelia.saalfeldlab.paintera.ui.source.state

import bdv.viewer.Source
import javafx.css.PseudoClass
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.layout.*
import javafx.scene.paint.Color
import org.janelia.saalfeldlab.fx.TextFields
import org.janelia.saalfeldlab.fx.extensions.TitledPaneExtensions
import org.janelia.saalfeldlab.fx.extensions.createNonNullValueBinding
import org.janelia.saalfeldlab.fx.extensions.createNullableValueBinding
import org.janelia.saalfeldlab.fx.extensions.nonnull
import org.janelia.saalfeldlab.fx.ui.NamedNode
import org.janelia.saalfeldlab.paintera.Style
import org.janelia.saalfeldlab.paintera.addStyleClass
import org.janelia.saalfeldlab.paintera.state.SourceState
import org.slf4j.LoggerFactory
import java.lang.invoke.MethodHandles
import java.util.function.Consumer

class StatePane(
	private val state: SourceState<*, *>,
	activeSourceRadioButtonGroup: ToggleGroup,
	remove: Consumer<Source<*>>,
) {

	private val nameProperty = state.nameProperty()

	private val statePaneVisibleProperty = state.isVisibleProperty
	var statePaneIsVisible: Boolean by statePaneVisibleProperty.nonnull()

	val pane = TitledPane(null, state.preferencePaneNode()).apply {
		HBox.setHgrow(this, Priority.ALWAYS)
		VBox.setVgrow(this, Priority.ALWAYS)
		isExpanded = false
		alignment = Pos.CENTER_RIGHT
		LOG.debug("_pane width is {} ({})", this.width, width)
	}

	init {
		val closeButton = Button(null).apply {
			addStyleClass("source-control", *Style.CLOSE_ICON.classes)
			onAction = EventHandler { remove.accept(state.dataSource) }
			tooltip = Tooltip("Remove source")
		}
		val activeSource = RadioButton().apply {
			tooltip = Tooltip("Select as active source")
			toggleGroup = activeSourceRadioButtonGroup
			userData = state.dataSource
		}

		val visibilityButton = Button(null).apply {
			addStyleClass("source-control", *Style.VISIBILITY_ICON.classes)
			statePaneVisibleProperty.subscribe { visible ->
				pseudoClassStateChanged(PseudoClass.getPseudoClass("visible"), visible)
			}
			onAction = EventHandler { statePaneIsVisible = !statePaneIsVisible }
			maxWidth = 20.0
			tooltip = Tooltip("Toggle visibility")
		}
		val nameField = TextFields.editableOnDoubleClick().apply {
			textProperty().bindBidirectional(nameProperty)
			tooltip = Tooltip().also {
				it.textProperty()
					.bind(nameProperty.createNullableValueBinding { name -> "Source $name: Double click to change name, enter to confirm, escape to discard." })
			}
			backgroundProperty().bind(editableProperty().createNonNullValueBinding { if (it) EDITABLE_BACKGROUND else UNEDITABLE_BACKGROUND })
			HBox.setHgrow(this, Priority.ALWAYS)
		}
		val titleBox = HBox(
			nameField,
			NamedNode.bufferNode(),
			activeSource,
			visibilityButton,
			closeButton
		).apply {
			alignment = Pos.CENTER
			padding = Insets(0.0, RIGHT_PADDING, 0.0, LEFT_PADDING)
		}
		with(TitledPaneExtensions) {
			pane.graphicsOnly(titleBox)
		}
		// TODO how to get underlined in TextField?
//        nameField.underlineProperty().bind(_isCurrentSource)

	}

	companion object {

		private val LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())

		private const val LEFT_PADDING = 0.0

		private const val RIGHT_PADDING = 0.0

		private val EDITABLE_BACKGROUND = Background(BackgroundFill(Color.WHITE, CornerRadii.EMPTY, Insets(-1.4, 0.0, 1.0, 2.0)))

		private val UNEDITABLE_BACKGROUND = Background(BackgroundFill(Color.WHITE.deriveColor(0.0, 1.0, 1.0, 0.5), CornerRadii.EMPTY, Insets.EMPTY))

	}

}
